presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java

```diff
@@ -243,6 +246,9 @@ private SqlQueryExecution(
                     }
                 }
             }
+
+            // Optionally build and validate plan immediately, before execution begins
+            planFuture = isQueryExecutionFailFastEnabled(getSession()) ? createLogicalPlanAsync() : null;
```

Comment:
I'm wondering if we can just inline the `createLogicalPlanAsync` method, and instead, switch the executor based on the session property.

* If the flag is disabled, use `MoreExecutors.sameThreadExecutor()`
* If the flag is enabled, use a new executor which is injected into this class, and which we have metrics for.  Note we define the executor currently used [HERE](https://github.com/tdcmeehan/presto/blob/6072d373184ba802eb0a3fd6a4e0d211114c6956/presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java#L263-L265), we could define a new one with more restricted paralellism, and which we could add dedicated metrics for (using a `QueryExecutionMBean` as you see in the link).  One thing we'd need to figure out is how to start the planner in the current thread if the queue were full, without a race, but I think this should be doable.

Reply:
When flag is enabled, sure I can add a new executor

For when it's disabled, IIUC, using something like `MoreExecutors.directExecutor()` would cause the plan to be validated immediately in the constructor, which would change the default behavior - I don't think we want to do this. I could still inline `createLogicalPlanAsync()` if you want to make it more compact?

Reply:
That's a good point, we wouldn't want that!  I suppose it should be as you said, if the flag is enabled, submit it to the new executor with restricted parallelism.  When we get to line 470, we should try to cancel it without interrupting it--if the cancellation succeeded or if the feature is not enabled, do it in the current thread, otherwise, `get()` the future.

Reply:
So when you say try to cancel the future at line 470, is it just in case the ExecutorService hasn't started running the thread, and doing it the current thread would be quicker?

Reply:
Yes, both quicker, and also it won't be subject to the concurrency restraints of the executor it would be running under.

Reply:
also, let's make the executor service parallelism configurable.

Reply:
ok, good idea